### PR TITLE
Deprecate allow_lazy

### LINKED ISSUE #3161: .reduce() on a DataArray with Dask distributed immediately executes the preceding portions of the computational graph
#### MCVE Code Sample

`.mean()` on a `DataArray` pointing to a Dask array returns a Dask array-containing `DataArray` as expected:

```python
da = xr.DataArray(np.zeros((500, 500, 500))).chunk((100, 100, 100)).mean('dim_0')
da
```

```
<xarray.DataArray (dim_1: 500, dim_2: 500)>
dask.array<shape=(500, 500), dtype=float64, chunksize=(100, 100)>
Dimensions without coordinates: dim_1, dim_2
```

Calling `.compute()` on this result produces the expected result:

```python
da = xr.DataArray(np.zeros((500, 500, 500))).chunk((100, 100, 100)).mean('dim_0').compute()
```

```
<xarray.DataArray (dim_1: 500, dim_2: 500)>
array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])
Dimensions without coordinates: dim_1, dim_2
```

The `.reduce()` method immediately executes all of the previously queued computations leading up to the new reduce method before even calling the supplied function.

```python
def func(x, axis=None):
    return x

da = xr.DataArray(np.zeros((500, 500, 500))).chunk((100, 100, 100)).mean('dim_0').reduce(func)
```

```
<xarray.DataArray (dim_1: 500, dim_2: 500)>
array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])
Dimensions without coordinates: dim_1, dim_2
```

#### Expected Output

A Dask array when `.reduce(func)` isn't followed up by `.compute()`.

#### Problem Description

When using Dask distributed, the computational graph you are constructing is immediately executed if you call `.reduce()` instead of adding that function as another node in the DAG. This graph execution happens before the function you pass to reduce is called.

#### Output of ``xr.show_versions()``
<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-693.21.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.4
libnetcdf: 4.6.2

xarray: 0.12.3
pandas: 0.25.0
numpy: 1.16.4
scipy: 1.3.0
netCDF4: 1.5.1.2
pydap: None
h5netcdf: None
h5py: 2.9.0
Nio: None
zarr: None
cftime: 1.0.3.4
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: 2.1.0
distributed: 2.1.0
matplotlib: 3.1.1
cartopy: None
seaborn: None
numbagg: None
setuptools: 41.0.1
pip: 19.2.1
conda: None
pytest: 5.0.1
IPython: 7.6.1
sphinx: 2.1.2
</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #3161 
 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by dcherian** [comment]
This is a breaking change though could also be considered a bug since we were computing dask arrays by default.

**Comment by dcherian** [comment]
@shoyer what do you think of this aggressive deprecation strategy (i.e.treating the current behaviour as a bug and deprecating explicit uses of `allow_lazy`)?

**Comment by shoyer** [comment]
Works for me!

**Comment by dcherian** [comment]
Now with tests!

ready for review & merge.

File: xarray/core/variable.py [hunk 1]
```diff
@@ -1466,7 +1467,17 @@ def reduce(
 
         if dim is not None:
             axis = self.get_axis_num(dim)
+
+        if allow_lazy is not None:
+            warnings.warn(
+                "allow_lazy will be deprecated in version 0.16.0. It is now True by default.",
```
Comment (mathause) [inline xarray/core/variable.py]:
```suggestion
                "allow_lazy is deprecated and will be removed in version 0.16.0. It is now True by default.",
```
?

**Comment by dcherian** [comment]
Thanks @mathause! I'll merge tomorrow if there are no comments.

**Comment by max-sixty** [comment]
LGTM!

**Pull request merged**